spikeinterface/sortingcomponents/dimensionality_reduction.py

```diff
@@ -0,0 +1,96 @@
+from sklearn.decomposition import IncrementalPCA
+from sklearn.preprocessing import FunctionTransformer
+from sklearn.pipeline import Pipeline
+
+from spikeinterface.sortingcomponents.peak_pipeline import run_peak_pipeline, PeakPipelineStep
+from spikeinterface.sortingcomponents.peak_detection import detect_peaks
+
+from joblib import dump, load
+
+class WaveformExtractorStep(PeakPipelineStep):
+    need_waveforms = True
+
+    def __init__(self, recording, ms_before=1., ms_after=1.,  peak_sign='neg', all_channels=True):
+        PeakPipelineStep.__init__(self, recording, ms_before=ms_before, ms_after=ms_after)
+        self.all_channels = all_channels
+        self.peak_sign = peak_sign
+        self._kwargs.update(dict(all_channels=all_channels, peak_sign=peak_sign))
+        self._dtype = recording.get_dtype()
+
+    def get_dtype(self):
+        return self._dtype
+
+    
+    def compute_buffer(self, traces, peaks, waveforms):
+        
+        return waveforms
+
+def stack_channel_with_spikes(waveform):
+    n_spikes, n_samples, n_channels = waveform.shape 
+
+    return waveform.transpose(0, 2, 1).reshape(n_spikes * n_channels, n_samples)
+
+def separate_channel_from_spikes(temporal_staggered_waveform, n_components, n_channels):
+    n_spikes = int(temporal_staggered_waveform.shape[0] / n_channels)
+    reshaped = temporal_staggered_waveform.reshape(n_spikes, n_channels, n_components)
+    return reshaped.transpose(0, 2, 1)
+
+def fit_temporal_pca(recording, n_components, job_kwargs, whiten=True):
```

Comment:
IMO, we can make full use of the `postprocessing` module here and this should be a `self` function:

```
from spikeinterface.postprocessing import compute_principal_components, get_template_channel_sparsity
from spikeinterface.sortingcomponents.peak_selection import select_peaks

peaks = detect_peaks(recording, method='by_channel',
                                       peak_sign='neg', detect_threshold=5, exclude_sweep_ms=0.1,
                                      **job_kwargs)
# we can either subsample peaks, or cut some shorter recording segments from the recording (let's say we go for the second option)
sub_peaks = select_peaks(peaks) 
# see https://github.com/SpikeInterface/spikeinterface/blob/master/spikeinterface/sortingcomponents/peak_selection.py#L8

# make a sorting object out of it:
sorting = si.NumpySorting.from_peaks(peaks, sampling_frequency=recording.sampling_frequency) 
# see https://github.com/SpikeInterface/spikeinterface/blob/master/spikeinterface/core/numpyextractors.py#L209

# extract waveforms fast and in-memory (provided that we don't have na enormous number of peaks)
we = si.extract_waveforms(recording, sorting, folder=None, mode="memory", **job_kwargs)
# in this case unit_ids are channel_ids

# compute sparsity based on local_radius_um (should be an input)
sparsity = get_template_channel_sparsity(we, method="radius", radius_um=local_radius_um)

# compute PCA by_channel_global (with sparsity)
pc = compute_principal_components(we, n_components=n_components, mode='by_channel_global', sparsity=sparsity)

pca_model = pc.get_pca_model()

# model_folder should be an input
pca_model_path = model_folder / "pca_model.pkl"

with pca_model_path.open("wb") as f:
      pickle.dump(pca_model, f)

self._kwargs["model_path"] = model_path
```

The saved model can be used as is in the `compute_buffer()` function.

Reply:
I have something similar now in the last commit.